test/enginetest.c

```diff
@@ -162,8 +162,8 @@ static int test_engines(void)
         ENGINE_free(ptr);
     }
     for (loop = 0; loop < NUMTOADD; loop++) {
-        OPENSSL_free((void *)ENGINE_get_id(block[loop]));
-        OPENSSL_free((void *)ENGINE_get_name(block[loop]));
+        OPENSSL_free((void *)(intptr_t)ENGINE_get_id(block[loop]));
+        OPENSSL_free((void *)(intptr_t)ENGINE_get_name(block[loop]));
```

Comment:
This is how modern systems discard const qualifiers without warnings, just `(void *)` alone does not eliminate warnings.  However `intptr_t` is C99, and optional.  We don't yet use `intptr_t` anywhere in OpenSSL, but we do use `uintptr_t` in `crypto/dso/dso_dlfcn.c` and `crypto/ppccap.c`.  If there's a need to run tests on platforms with no `intptr_t` we could define a non-public `OPENSSL_UNCONST` macro instead, and make it be `(void *)` on some older platforms, and otherwise use `uintptr_t` or `intptr_t` which seems more portable.

Reply:
I'd argue that first sentence is a bit misleading. In sense that as far as standard goes `(void *)` alone is sufficient. And the warning is rather developer's choice (by means of additional -W compiler flag) than the way modern systems go about doing this kind of casts. And this calls for following question. The interface is obviously poorly designed. In sense that subroutines in question shouldn't return const. Of if they do, then engine layer should free the strings in question. Either way, it's only appropriate to do something about interface. Rather than swiping the problem under the rug with more elaborate cast that is. We do explicitly suppress some warnings as excessive, so one can as well do same for this one, -Wcast-qual...

Reply:
Yes, and the reason for all the inappropriate use of const in the code (lack of const or other const cast-aways) is because const compiler warnings are elided. I suggest instead to turn the warnings on, and provide a macro or other way to silence the ones like the above...

Reply:
Do note that I'm not disputing *other* casts(*), only this specific pair. What I'm saying is that insisting on elaborate warning masking in *this* particular case is equivalent of swiping the problem of bad interface design under the carpet. And that it's a bad thing. And that we would be better off addressing the interface design. And that if the design problem is not resolved on the spot we would be better off not engaging in elaborate casting, but rather put -Wcast-qual on todo list.

(*) Though some of others are not actually required. Most notably one doesn't actually have to pass const pointer to `foo(const void *p)`. const qualifier in argument is merely a "contract" that foo won't modify the object, it puts no obligation on caller to pass const-qualified pointer.

Reply:
We are in full agreement, fixing the interface to provide a way to free the data is the better way compared to silencing the compiler.